src/Disks/ObjectStorages/S3/diskSettings.cpp

```diff
@@ -92,19 +103,27 @@ std::unique_ptr<S3::Client> getClient(
 
     const Settings & local_settings = context->getSettingsRef();
 
-    int s3_max_redirects = static_cast<int>(global_settings[Setting::s3_max_redirects]);
+    unsigned int s3_max_redirects = static_cast<unsigned int>(request_settings[S3RequestSetting::max_redirects]);
```

Comment:
We create Client only once at the start. New configs are not applied to the client instance.
Am not sure that it is possible to achieve here that
> make settings of S3 disk can be hot reload if change the config XML file

Reply:
If we modify the settings in `<disks>` section, the config reloader will call 

https://github.com/ClickHouse/ClickHouse/blob/e8935e464f8737a11c2d8c4d3caddb305499dd1c/src/Disks/IDisk.h#L488

to apply the new disk settings.

And it will eventually call 

https://github.com/ClickHouse/ClickHouse/blob/e8935e464f8737a11c2d8c4d3caddb305499dd1c/src/Disks/ObjectStorages/S3/S3ObjectStorage.cpp#L484-L516

to replace the s3 client in a copy-on-write pattern.

Reply:
I see.
You do here a good thing. You put some set of settings inside the 'disks' section. Right now it just placed in profile section.

As for me, I would be glad if you totally remove that settings from profile section. They do not fit in it. Defaults should be places in server settings. Overrides should be in 'disks' section. So if you remove them from profile section and place as server settings, it would be more neat.